ecosystem/typescript/sdk_v2/src/bcs/serializer.ts

```diff
@@ -0,0 +1,246 @@
+   * serializer.serializeU16(4660);
+   * assert(serializer.getBytes() === new Uint8Array([0x34, 0x12]));
+   * ```
+   */
+  @checkNumberRange(0, MAX_U16_NUMBER)
+  serializeU16(value: Uint16) {
+    this.serializeWithFunction(DataView.prototype.setUint16, 2, value);
+  }
+
+  /**
+   * Serializes a uint32 number.
+   *
+   * BCS layout for "uint32": Four bytes. Binary format in little-endian representation.
+   * @example
+   * ```ts
+   * const serializer = new Serializer();
+   * serializer.serializeU32(305419896);
+   * assert(serializer.getBytes() === new Uint8Array([0x78, 0x56, 0x34, 0x12]));
+   * ```
+   */
+  @checkNumberRange(0, MAX_U32_NUMBER)
+  serializeU32(value: Uint32) {
+    this.serializeWithFunction(DataView.prototype.setUint32, 4, value);
+  }
+
+  /**
+   * Serializes a uint64 number.
+   *
+   * BCS layout for "uint64": Eight bytes. Binary format in little-endian representation.
+   * @example
+   * ```ts
+   * const serializer = new Serializer();
+   * serializer.serializeU64(1311768467750121216);
+   * assert(serializer.getBytes() === new Uint8Array([0x00, 0xEF, 0xCD, 0xAB, 0x78, 0x56, 0x34, 0x12]));
+   * ```
+   */
+  @checkNumberRange(BigInt(0), MAX_U64_BIG_INT)
+  serializeU64(value: AnyNumber) {
+    const low = BigInt(value.toString()) & BigInt(MAX_U32_NUMBER);
+    const high = BigInt(value.toString()) >> BigInt(32);
+
+    // write little endian number
+    this.serializeU32(Number(low));
+    this.serializeU32(Number(high));
+  }
+
+  /**
+   * Serializes a uint128 number.
+   *
+   * BCS layout for "uint128": Sixteen bytes. Binary format in little-endian representation.
+   */
+  @checkNumberRange(BigInt(0), MAX_U128_BIG_INT)
+  serializeU128(value: AnyNumber) {
+    const low = BigInt(value.toString()) & MAX_U64_BIG_INT;
+    const high = BigInt(value.toString()) >> BigInt(64);
+
+    // write little endian number
+    this.serializeU64(low);
+    this.serializeU64(high);
+  }
+
+  /**
+   * Serializes a uint256 number.
+   *
+   * BCS layout for "uint256": Sixteen bytes. Binary format in little-endian representation.
+   */
+  @checkNumberRange(BigInt(0), MAX_U256_BIG_INT)
+  serializeU256(value: AnyNumber) {
+    const low = BigInt(value.toString()) & MAX_U128_BIG_INT;
+    const high = BigInt(value.toString()) >> BigInt(128);
+
+    // write little endian number
+    this.serializeU128(low);
+    this.serializeU128(high);
+  }
+
+  /**
+   * Serializes a uint32 number with uleb128.
+   *
+   * BCS uses uleb128 encoding in two cases: (1) lengths of variable-length sequences and (2) tags of enum values
+   */
+  @checkNumberRange(0, MAX_U32_NUMBER)
+  serializeU32AsUleb128(val: Uint32) {
+    let value = val;
+    const valueArray = [];
+    while (value >>> 7 !== 0) {
+      valueArray.push((value & 0x7f) | 0x80);
+      value >>>= 7;
+    }
+    valueArray.push(value);
+    this.appendToBuffer(new Uint8Array(valueArray));
+  }
+
+  /**
+   * Returns the buffered bytes
+   */
+  getBytes(): Uint8Array {
```

Comment:
Should we rename this to `toUint8Array`?

Reply:
yeah good idea! Could be much clearer